### DISCUSSION TITLE
Dynamic imports with query parameters

### DISCUSSION THREAD

**Discussion by youroff** [webpack/webpack#13774] [General]
## Feature request

**What is the expected behavior?**
I'm trying to use dynamic module loading like this:
```
  const App = lazy(() => import('./protected?token=' + token));

  return <Suspense fallback={<div>Loading... </div>}>
    {token && <App/> || <Login/>}  
  </Suspense>;
```
Currently the module is perfectly resolved and split when no query parameters used. Once I add query param, the build fails with: `protected\:5 Uncaught Error: Cannot find module './protected?token=XXXXXXX'`

**What is motivation or use case for adding/changing the behavior?**
In my case it's some authentication for the part of the app code.

**How should this be implemented in your opinion?**
`?` and the rest could be just stripped on the module resolution and added back once the final url is formed.

**Are you willing to work on this yourself?**
yes, with some hints

**Comment by alexander-akait** [comment]
> Currently the module is perfectly resolved and split when no query parameters used. Once I add query param, the build fails with: protected\:5 Uncaught Error: Cannot find module './protected?token=XXXXXXX'


It means you have invalid URL, please create reproducible example, so I can help you

**Reply by alexander-akait** [reply]
Note your should use URL in `import`, so valid code is ` const App = lazy(() => import('./protected.js?token=' + token));
`

**Reply by youroff** [reply]
I tried that, didn't work. Still getting: `Uncaught Error: Cannot find module './protected.js?token=asfasfas'`

**Reply by youroff** [reply]
Even without a token... this works fine `import('./protected')`, this doesn't: `import('./protected.js')`

**Reply by alexander-akait** [reply]
Do you have `protected.js`? If no I think you need https://github.com/webpack/webpack/discussions/13738

**Reply by youroff** [reply]
I have `protected/index.tsx` which I think resolved by webpack into some dynamic chunk. I tried more manual approach: made a separate entrypoint for protected portion of the app. Then used dynamic import with `webpackIgnore: true`, but then I couldn't make that chunk a module, so it wouldn't load. Apparently `chunkFormat` option `module` is not implemented (yet?)...

**Reply by alexander-akait** [reply]
> Apparently chunkFormat option module is not implemented (yet?)...

https://github.com/webpack/webpack/blob/main/lib/WebpackOptionsApply.js#L143, implemented in the latest version

**Reply by youroff** [reply]
Oh nice, I'll try that... Would it apply to both chunks though? Cause I think I need only one chunk as a module.

**Reply by alexander-akait** [reply]
only for whole bundle

**Reply by youroff** [reply]
Would 2 configs passed as an array solve it?

**Reply by alexander-akait** [reply]
Yes, should work

**Reply by youroff** [reply]
So... maybe I'm doing something wrong, but even with `chunkFormat: "module"` I still get the file that starts with `!function(){"use strict";var t=...` so certainly not a module:

**Comment by youroff** [comment]
```
const rules = [
  {
    test: /\.(ts|js)x?$/i,
    exclude: /node_modules/,
    use: {
      loader: "babel-loader",
      options: {
        presets: [
          "@babel/preset-env",
          "@babel/preset-react",
          "@babel/preset-typescript",
        ],
      },
    },
  },
  {
    test: /\.css$/i,
    use: ["style-loader", "css-loader", "postcss-loader"],
  },
  {
    test: /\.(png|jpe?g|gif)$/i,
    use: [
      {
        loader: 'file-loader',
      },
    ],
  }
];

const main = {
  mode: "production",
  optimization: {
    usedExports: true
  },
  entry: "./src/index.tsx",
  output: {
    filename: "[name].[contenthash].js",
    path: path.resolve(__dirname, "build"),
    publicPath: "",
  },
  module: {
    rules: rules,
  },
  resolve: {
    extensions: [".tsx", ".ts", ".js"],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: "[name].[contenthash].css",
    }),
    new HtmlWebpackPlugin({
      template: "src/index.html",
      favicon: "public/favicon.ico"
    }),
    new CompressionPlugin(),
  ]
};

const protected = {
  mode: "production",
  optimization: {
    usedExports: true
  },
  entry: "./src/protected/index.tsx",
  output: {
    path: path.resolve(__dirname, "build"),
    chunkFormat: "module",
    filename: "protected/app.js",
    publicPath: "",
  },
  module: {
    rules: rules
  },
  resolve: {
    extensions: [".tsx", ".ts", ".js"],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: "[name].[contenthash].css",
    }),
    new CleanWebpackPlugin.CleanWebpackPlugin(),
    new CompressionPlugin()
  ]
};

module.exports = [main, protected];
```

**Reply by alexander-akait** [reply]
You need enable https://webpack.js.org/configuration/experiments/#experimentsoutputmodule, anyway I don't undestand why you try these dirty workarounds, please create reproducible test repo

**Reply by youroff** [reply]
Should I inform babel somehow that it should leave it as module?

**Reply by alexander-akait** [reply]
Ideally you should never change `import` on `require` for babel

**Reply by youroff** [reply]
I'm trying these dirty workarounds because I need to dynamically load part of application with a token and straightforward way breaks when I add a token. Sure, I'll create a repo in a bit. Thanks!

**Reply by youroff** [reply]
Here you go: https://github.com/youroff/module_loading_repro

**Reply by youroff** [reply]
Ugh... some of ts setup leaked there, but it shouldn't matter

**Reply by youroff** [reply]
Hey @alexander-akait did you have a chance to take a look at the reproduction repo? I wonder if it should be filed as an issue, since it doesn't work as expected?

**Reply by alexander-akait** [reply]
I my roadmap, try to find time today/tomorrow, sorry for delay

**Reply by youroff** [reply]
Thank you, take your time and let me know if I can help in any way! :)

**Reply by alexander-akait** [reply]
Due usage `import('./module.js?token=' + token)`, import is not static analyzable, why you need this? If you need dynamic import you can use, you can use:
```js
output: {
  publicPath: "/",
  chunkFilename: "[name].js?token=something"
},
```

Also we support `Function`, i.e.
```js
{
  chunkFilename: (data) => {
    console.log(data.chunk.id);
      
    return "[name].js?token=something"
  }
}
```

**Reply by youroff** [reply]
I need it to put the app code behind the authentication. For this reason using a token in webpack config won't work, since it's know at runtime and constantly changes. But there might be less crazy use cases and I've seen people on the Internet having similar issues while trying to dynamic-load some configurable external assets (in which case they could just use webpackIgnore).

**Reply by alexander-akait** [reply]
So you store token in runtime code?

**Reply by youroff** [reply]
I don't, I'm using dynamic import semantics, the token is coming into this function from elsewhere.

**Reply by alexander-akait** [reply]
in this case maybe better to create plugin which push logic to get token in webpack runtime and add query to necessary requests, here you can look at hook https://github.com/webpack/webpack/blob/main/lib/RuntimePlugin.js

**Reply by youroff** [reply]
Yeah, I was considering writing a plugin for this, I'm just not familiar with webpack internals, so it might take a while. Would you be interested in contracting for us? If yes, how can I reach you?

**Reply by alexander-akait** [reply]
Hello, sorry I am unavailable for this job, a lot of own tasks :disappointed:

**Reply by youroff** [reply]
Ok, no problem. Thanks for your help anyways! ;)
